paywall/src/unlock.js/MainWindowHandler.ts

```diff
@@ -0,0 +1,266 @@
+import {
+  LockStatus,
+  UnlockWindowNoProtocolYet,
+  UnlockWindow,
+} from '../windowTypes'
+import IframeHandler from './IframeHandler'
+import { PostMessages } from '../messageTypes'
+import { PaywallConfig } from '../unlockTypes'
+
+interface hasPrototype {
+  prototype?: any
+}
+
+/**
+ * This class handles all of the messaging to the window object
+ *
+ * Specifically, sending the "unlockProtocol" event, creating the
+ * "unlockProtocol" object on window, and also all of the showing
+ * and hiding of checkout and account iframes
+ */
+export default class MainWindowHandler {
+  private window: UnlockWindowNoProtocolYet
+  private iframes: IframeHandler
+  private showCheckoutWhenAccountsHides: boolean = false
+  private showingCheckout: boolean = false
+  private showingAccountsIframe: boolean = false
+  private lockStatus: LockStatus = undefined
+  private config: PaywallConfig
+
+  constructor(
+    window: UnlockWindowNoProtocolYet,
+    iframes: IframeHandler,
+    config: PaywallConfig
+  ) {
+    this.window = window
+    this.iframes = iframes
+    this.config = config
+  }
+
+  init() {
+    // create window.unlockProtocol
+    this.setupUnlockProtocolVariable()
+
+    // this is a cache for the time between script startup and the full load
+    // of the data iframe. The data iframe will then send down the current
+    // value, overriding this. A bit later, the blockchain handler will update
+    // with the actual value, so this is only used for a few milliseconds
+    const locked = this.getCachedLockState()
+    if (locked === true) {
+      this.dispatchEvent('locked')
+    }
+    if (locked === false) {
+      this.dispatchEvent('unlocked')
+    }
+
+    // respond to "unlocked" and "locked" events by
+    // dispatching "unlockProtocol" on the main window
+    // and
+    this.iframes.data.on(PostMessages.LOCKED, () => {
+      this.dispatchEvent('locked')
+      this.setCachedLockedState(true)
+      // Update the user-facing status with locked/unlocked updates
+      this.lockStatus = 'locked'
+    })
+    this.iframes.data.on(PostMessages.UNLOCKED, () => {
+      this.dispatchEvent('unlocked')
+      this.setCachedLockedState(false)
+      // Update the user-facing status with locked/unlocked updates
+      this.lockStatus = 'unlocked'
+    })
+
+    // handle display of checkout and account UI
+    this.iframes.checkout.on(PostMessages.DISMISS_CHECKOUT, () => {
+      this.hideCheckoutIframe()
+    })
+
+    this.iframes.checkout.on(PostMessages.READY, () => {
```

Comment:
This should probably go away as we're deprecating the "paywall" type altogether. Do you mind adding a comment?